s2tbx-s2msi-reader/src/main/java/org/esa/s2tbx/dataio/s2/Sentinel2ProductReader.java

```diff
@@ -234,55 +246,117 @@ private Path getQuickLookFile(VirtualPath inputVirtualPath) throws IOException {
+                    + ", columnCount=" + mosaicMatrix.getColumnCount() + ".");
         }
     }
 
     protected static int computeMatrixCellsResolutionCount(MosaicMatrix mosaicMatrix) {
+        return computeMatrixCellsResolutionCount(mosaicMatrix, false);
+    }
+
+    protected static int computeMatrixCellsResolutionCount(MosaicMatrix mosaicMatrix, boolean geoTiffOption) {
         if (mosaicMatrix.getRowCount() > 0 && mosaicMatrix.getColumnCount() > 0) {
-            JP2MosaicBandMatrixCell firstMatrixCell = (JP2MosaicBandMatrixCell)mosaicMatrix.getCellAt(0, 0);
-            for (int rowIndex = 0; rowIndex < mosaicMatrix.getRowCount(); rowIndex++) {
-                for (int columnIndex = 0; columnIndex < mosaicMatrix.getColumnCount(); columnIndex++) {
-                    JP2MosaicBandMatrixCell matrixCell = (JP2MosaicBandMatrixCell)mosaicMatrix.getCellAt(rowIndex, columnIndex);
-                    if (firstMatrixCell.getResolutionCount() != matrixCell.getResolutionCount()) {
-                        throw new IllegalStateException("Different resolution count: cell at "+rowIndex+", "+columnIndex+" has data type " + matrixCell.getResolutionCount()+" and cell at "+0+", "+0+" has resolution count " + firstMatrixCell.getResolutionCount()+".");
-                    }
+            if (geoTiffOption)
+                return computeTiffMatrixCellsResolutionCount(mosaicMatrix);
+            else
+                return computeJP2MatrixCellsResolutionCount(mosaicMatrix);
+        } else {
+            throw new IllegalArgumentException("The matrix is empty: rowCount=" + mosaicMatrix.getRowCount()
+                    + ", columnCount=" + mosaicMatrix.getColumnCount() + ".");
+        }
+    }
+
+    protected static int computeJP2MatrixCellsResolutionCount(MosaicMatrix mosaicMatrix) {
+        JP2MosaicBandMatrixCell firstMatrixCell = (JP2MosaicBandMatrixCell) mosaicMatrix.getCellAt(0, 0);
+        for (int rowIndex = 0; rowIndex < mosaicMatrix.getRowCount(); rowIndex++) {
+            for (int columnIndex = 0; columnIndex < mosaicMatrix.getColumnCount(); columnIndex++) {
+                JP2MosaicBandMatrixCell matrixCell = (JP2MosaicBandMatrixCell) mosaicMatrix.getCellAt(rowIndex,
+                        columnIndex);
+                if (firstMatrixCell.getResolutionCount() != matrixCell.getResolutionCount()) {
+                    throw new IllegalStateException("Different resolution count: cell at " + rowIndex + ", "
+                            + columnIndex + " has data type " + matrixCell.getResolutionCount() + " and cell at " + 0
+                            + ", " + 0 + " has resolution count " + firstMatrixCell.getResolutionCount() + ".");
+                }
+            }
+        }
+        return firstMatrixCell.getResolutionCount();
+    }
+
+    protected static int computeTiffMatrixCellsResolutionCount(MosaicMatrix mosaicMatrix) {
+        GeoTiffMatrixCell firstMatrixCell = (GeoTiffMatrixCell) mosaicMatrix.getCellAt(0, 0);
```

Comment:
@oanahogoiu @kraftek 
Do you know about this method? I don't retrieve the correct resolution count with the tiff images (it's always equal one).